File: backend/services/delivery/internal/entity/delivery.go
package entity

import "time"

// DeliveryStatus is the lifecycle state of a delivery record.
type DeliveryStatus string

const (
	DeliveryAvailable      DeliveryStatus = "AVAILABLE"
	DeliveryClaimed        DeliveryStatus = "CLAIMED"
	DeliveryPickedUp       DeliveryStatus = "PICKED_UP"
	DeliveryDelivering     DeliveryStatus = "DELIVERING"
	DeliveryDelivered      DeliveryStatus = "DELIVERED"
	DeliveryCancelled      DeliveryStatus = "CANCELLED"
	DeliveryStoreDelivering DeliveryStatus = "STORE_DELIVERING"
)

// Delivery tracks a single order's delivery lifecycle.
// order_id is UNIQUE — one delivery record per order.
type Delivery struct {
	ID          string         `gorm:"type:uuid;primaryKey;default:gen_random_uuid()"`
	OrderID     string         `gorm:"type:uuid;not null;uniqueIndex"`
	StoreID     string         `gorm:"type:uuid;not null;index"`
	LocationID  string         `gorm:"type:uuid;not null"`
	CustomerID  string         `gorm:"type:uuid;not null"`
	ShipperID   *string        `gorm:"type:uuid;index"`
	ShipFee     int64          `gorm:"not null;default:0"`
	Status      DeliveryStatus `gorm:"type:varchar(20);not null;default:'AVAILABLE';index"`
	BatchID     *string        `gorm:"type:uuid"`
	ClaimedAt   *time.Time     `gorm:"type:timestamptz"`
	DeliveredAt *time.Time     `gorm:"type:timestamptz"`
	CreatedAt   time.Time      `gorm:"type:timestamptz;not null;default:now()"`
	UpdatedAt   time.Time      `gorm:"type:timestamptz;not null;default:now()"`
}

func (Delivery) TableName() string { return "deliveries" }

// ValidDeliveryTransitions defines allowed shipper status transitions.
var ValidDeliveryTransitions = map[DeliveryStatus]DeliveryStatus{
	DeliveryClaimed:    DeliveryPickedUp,
	DeliveryPickedUp:   DeliveryDelivering,
	DeliveryDelivering: DeliveryDelivered,
}

// CanTransitionTo reports whether a shipper-driven transition from current to next is valid.
func (d *Delivery) CanTransitionTo(next DeliveryStatus) bool {
	allowed, ok := ValidDeliveryTransitions[d.Status]
	return ok && allowed == next
}

File: frontend/src/features/deliveries/types/delivery.ts
// Mirrors backend delivery-service entities.
// Entity fields are PascalCase (Go JSON default). Request bodies use snake_case.

export type DeliveryStatus =
  | 'AVAILABLE'
  | 'CLAIMED'
  | 'PICKED_UP'
  | 'DELIVERING'
  | 'DELIVERED'
  | 'CANCELLED'
  | 'STORE_DELIVERING';

// GET /api/v1/deliveries/available — one item in the pool
export interface AvailableDelivery {
  order_id: string;
  order_code?: string;
  store_id: string;
  location_id: string;
  room_path: string; // resolved by grpcclient/location; fallback = raw UUID
  ship_fee: number;
  desired_time?: string; // RFC3339
  created_at: string;
}

// Delivery record from GET /api/v1/me/deliveries .deliveries[] — snake_case DTO.
export interface MyDelivery {
  id: string;
  order_id: string;
  order_code?: string;
  store_id: string;
  location_id: string;
  location_level?: string;
  ship_fee: number;
  status: DeliveryStatus;
  batch_id?: string;
  claimed_at?: string;
  delivered_at?: string;
  desired_time?: string;    // RFC3339 — customer's requested receive time
  late_by_minutes?: number; // > 0 means delivery was late
  created_at: string;
}

// GET /api/v1/me/deliveries response data shape
export interface MyDeliveriesData {
  deliveries: MyDelivery[];
  earnings: number;
}

// POST /api/v1/deliveries/:orderId/claim → 201
export interface ClaimResult {
  batch_id: string;
}

// PATCH /api/v1/deliveries/:orderId/status body
export interface UpdateStatusBody {
  status: 'PICKED_UP' | 'DELIVERING' | 'DELIVERED';
}

// POST /api/v1/deliveries/:orderId/incident body
export interface ReportIncidentBody {
  type: string;
  note: string;
  photo_url?: string;
}

// Incident types for the dropdown
export const INCIDENT_TYPES = [
  { value: 'WRONG_ADDRESS', label: 'Sai địa chỉ' },
  { value: 'CUSTOMER_ABSENT', label: 'Khách không có mặt' },
  { value: 'ITEM_DAMAGED', label: 'Hàng bị hỏng' },
  { value: 'OTHER', label: 'Khác' },
] as const;

// Map an incident type code → Vietnamese label (falls back to the raw code).
export function incidentTypeLabel(type: string): string {
  return INCIDENT_TYPES.find((t) => t.value === type)?.label ?? type;
}

// GET /api/v1/admin/incidents — admin-facing incident (resolved order code + shipper name).
export interface AdminIncident {
  id: string;
  order_code: string;
  shipper_name: string;
  type: string;
  note: string;
  photo_url?: string | null;
  status: 'OPEN' | 'RESOLVED';
  created_at: string;
}
